core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -270,26 +270,38 @@ class LogManager(logDirs: Seq[File],
   def directoryId(dir: String): Option[Uuid] = dirIds.get(dir)
 
   /**
-   * Determine directory ID for each directory with a meta.properties.
+   * Determine directory ID for each directory.
    * If meta.properties does not include a directory ID, one is generated and persisted back to meta.properties.
-   * Directories without a meta.properties don't get a directory ID assigned.
+   * Directories without a meta.properties file, which can only occur in Zk mode, will file have a directory ID assigned.
+   * The ID will be written when the new meta.properties file is first written.
    */
   private def directoryIds(dirs: Seq[File]): Map[String, Uuid] = {
+    val s = scala.collection.mutable.Set[Uuid]()
     dirs.flatMap { dir =>
       try {
         val metadataCheckpoint = new BrokerMetadataCheckpoint(new File(dir, KafkaServer.brokerMetaPropsFile))
-        metadataCheckpoint.read().map { props =>
-          val rawMetaProperties = new RawMetaProperties(props)
-          val uuid = rawMetaProperties.directoryId match {
-            case Some(uuidStr) => Uuid.fromString(uuidStr)
-            case None =>
-              val uuid = Uuid.randomUuid()
-              rawMetaProperties.directoryId = uuid.toString
-              metadataCheckpoint.write(rawMetaProperties.props)
-              uuid
+        val uuid = metadataCheckpoint.read() match {
+          case Some(props) => {
+            val rawMetaProperties = new RawMetaProperties(props)
+            val uuid_from_properties = rawMetaProperties.directoryId match {
+              case Some(uuidStr) => Uuid.fromString(uuidStr)
+              case None =>
+                val uuid_new = Uuid.randomUuid()
+                rawMetaProperties.directoryId = uuid_new.toString
+                metadataCheckpoint.write(rawMetaProperties.props)
+                uuid_new
+            }
+            uuid_from_properties
+          }
+          case None => {
+            Uuid.randomUuid()
           }
-          dir.getAbsolutePath -> uuid
-        }.toMap
+        }
+        if (s.contains(uuid)) {
+            throw new RuntimeException(s"Found duplicate directory.ids ${uuid.toString}")
+        }
```

Comment:
An alternative to maintaining a mutable set and modifying it from within the flatMap lamda could be to compare the size of the result with `.toSet.size` or `.toSeq.distinct.size` e.g.

```scala
if (dirIds.values.toSet.size != dirIds.values.size)
  throw new RuntimeException(s"Found duplicate directory.ids ${uuid.toString}")
```

Reply:
Perhaps `KafkaException` is more suitable than `RuntimeException`?